### PR TITLE
25 as admin i can get messages sent by visitors dla juniora

### PR DESCRIPTION
**Pull request by KrystynaCal**

### CODE REVIEW

File: src/main/java/karuna/karuna_backend/visitor/message/domain/VisitorMessageService.java [hunk 1]
```diff
@@ -15,4 +19,18 @@ public VisitorMessageDto sendMessage(VisitorMessageCreateDto visitorMessageCreat
...
+
+    public List<VisitorMessageDto> getMessages(VisitorMessageRequest visitorMessageRequest) {
+        PageRequest pageRequest = PageRequest.of(visitorMessageRequest.offset(), visitorMessageRequest.limit());
+        List<VisitorMessage> messages = visitorMessageRepository.findAllByOrderByCreatedAtDesc(pageRequest);
+        messages.forEach(message -> {
+            if (message.getBody().length() > visitorMessageRequest.bodyLenLimit()) {
+                message.setBody(message.getBody().substring(0, visitorMessageRequest.bodyLenLimit()));
+            }
+        });
+        List<VisitorMessageDto> messagedtos = messages.stream()
```
Comment (misterrevo) [inline src/main/java/karuna/karuna_backend/visitor/message/domain/VisitorMessageService.java:32]: U don't need to declare variable if u do nothing with it, just return

**Review by misterrevo — COMMENTED** [review]
@KrystynaCal Write tests with new convention

**Comment by KrystynaCal** [comment]
I will add tests when the code passes review

File: src/main/java/karuna/karuna_backend/visitor/message/domain/VisitorMessageRepository.java [hunk 2]
```diff
@@ -1,9 +1,11 @@
 package karuna.karuna_backend.visitor.message.domain;
 
 import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.stereotype.Repository;
 
 import java.util.List;
 
+@Repository
```
Comment (misterrevo) [inline src/main/java/karuna/karuna_backend/visitor/message/domain/VisitorMessageRepository.java:9]: Is unused when u extends jparepository

Comment (KrystynaCal) [reply src/main/java/karuna/karuna_backend/visitor/message/domain/VisitorMessageRepository.java:9]: according to discuss on discord I leave it as it is

File: src/main/java/karuna/karuna_backend/visitor/message/domain/VisitorMessageService.java [hunk 3]
```diff
@@ -15,4 +20,17 @@ public VisitorMessageDto sendMessage(VisitorMessageCreateDto visitorMessageCreat
         VisitorMessage visitorMessage = VisitorMessageMapper.toModel(visitorMessageCreateDto);
         return VisitorMessageMapper.toDto(visitorMessageRepository.save(visitorMessage));
     }
+
+
+    public VisitorMessageDtoList getMessages(VisitorMessageRequest visitorMessageRequest) {
```
Comment (misterrevo) [inline src/main/java/karuna/karuna_backend/visitor/message/domain/VisitorMessageService.java:25]: VistiroMessageWrapper not will be better?

File: src/main/java/karuna/karuna_backend/visitor/message/domain/VisitorMessageService.java [hunk 4]
```diff
@@ -15,4 +20,17 @@ public VisitorMessageDto sendMessage(VisitorMessageCreateDto visitorMessageCreat
...
+        PageRequest pageRequest = PageRequest.of(visitorMessageRequest.offset(), visitorMessageRequest.limit());
+        List<VisitorMessage> messages = visitorMessageRepository.findAllByOrderByCreatedAtDesc(pageRequest);
+        messages.forEach(message -> {
+            if (message.getBody().length() > visitorMessageRequest.bodyLenLimit()) {
+                message.setBody(message.getBody().substring(0, visitorMessageRequest.bodyLenLimit()));
+            }
+        });
+        return new VisitorMessageDtoList(messages.stream()
+                .map(VisitorMessageMapper::toDto).toList());
```
Comment (misterrevo) [inline src/main/java/karuna/karuna_backend/visitor/message/domain/VisitorMessageService.java:34]: Its cosmetic, but i always move single operation on stream to new line

File: src/main/java/karuna/karuna_backend/visitor/message/dto/VisitorMessageDtoList.java [hunk 5]
```diff
@@ -0,0 +1,6 @@
+package karuna.karuna_backend.visitor.message.dto;
+
+import java.util.List;
+
+public record VisitorMessageDtoList(List<VisitorMessageDto> messagesDto) {
```
Comment (misterrevo) [inline src/main/java/karuna/karuna_backend/visitor/message/dto/VisitorMessageDtoList.java:5]: I think in json i want to see message [] not messagesDto

File: src/main/java/karuna/karuna_backend/visitor/message/infrastructure/VisitorMessageController.java [hunk 6]
```diff
@@ -42,4 +47,17 @@ VisitorMessageDto sendMessage(@RequestBody @Valid VisitorMessageCreateDto visito
...
+
+
+    @Operation(summary = "Get messages", description = "Returns a list of messages with pagination and limited body length")
+    @ApiResponses({
+            @ApiResponse(
+                    responseCode = "200",
+                    description = "List od messages",
+                    content = @Content(mediaType = "application/json", schema = @Schema(implementation = VisitorMessageDtoList.class))),
+    })
+    @GetMapping
```
Comment (misterrevo) [inline src/main/java/karuna/karuna_backend/visitor/message/infrastructure/VisitorMessageController.java:59]: Add 403 to swagger documentation

File: src/main/java/karuna/karuna_backend/visitor/message/domain/VisitorMessageService.java (hunk 4 above)
Comment (misterrevo) [reply src/main/java/karuna/karuna_backend/visitor/message/domain/VisitorMessageService.java:34]: and mayby move it to mapper?

Comment (misterrevo) [reply src/main/java/karuna/karuna_backend/visitor/message/domain/VisitorMessageService.java:34]: Like toWrapper(List<VIsitroMessage> vistiorMessages)

File: src/main/java/karuna/karuna_backend/visitor/message/domain/VisitorMessageService.java (hunk 3 above)
Comment (KrystynaCal) [reply src/main/java/karuna/karuna_backend/visitor/message/domain/VisitorMessageService.java:29]: i will change it

File: src/main/java/karuna/karuna_backend/visitor/message/dto/VisitorMessageDtoList.java (hunk 5 above)
Comment (KrystynaCal) [reply src/main/java/karuna/karuna_backend/visitor/message/dto/VisitorMessageDtoList.java:5]: i will name it "messages"

File: src/main/java/karuna/karuna_backend/security/config/WebSecurityConfig.java [hunk 7]
```diff
@@ -42,11 +43,18 @@ SecurityFilterChain filterChain(HttpSecurity http) throws Exception {
                                 "/swagger-ui/**",
                                 "/swagger-ui.html",
                                 "/v3/api-docs/**",
-                                "/swagger-resources/**",
-                                "/visitors/messages/**")
+                                "/swagger-resources/**")
                         .permitAll()
+                        .requestMatchers(HttpMethod.POST, "/visitors/messages").hasAnyRole("USER", "ADMIN")
```
Comment (misterrevo) [inline src/main/java/karuna/karuna_backend/security/config/WebSecurityConfig.java:48]: Is not needed .anyRequest().authenticated() make the same thing, i write u on dc

Comment (misterrevo) [reply src/main/java/karuna/karuna_backend/security/config/WebSecurityConfig.java:48]: we dont have more roles than user and admin so authenticated is equal to be user or admin

File: src/main/java/karuna/karuna_backend/visitor/message/domain/VisitorMessageService.java [hunk 8]
```diff
@@ -19,4 +24,18 @@ public VisitorMessageDto sendMessage(VisitorMessageCreateDto visitorMessageCreat
...
+
+    public VisitorMessageWrapper getMessages(VisitorMessageRequest visitorMessageRequest) {
+        PageRequest pageRequest = PageRequest.of(visitorMessageRequest.offset(), visitorMessageRequest.limit());
+        List<VisitorMessage> messages = visitorMessageRepository.findAllByOrderByCreatedAtDesc(pageRequest);
+        messages.forEach(message -> {
+            if (message.getBody().length() > visitorMessageRequest.bodyLenLimit()) {
+                message.setBody(message.getBody().substring(0, visitorMessageRequest.bodyLenLimit()));
+            }
+        });
+        return new VisitorMessageWrapper(messages.stream()
```
Comment (misterrevo) [inline src/main/java/karuna/karuna_backend/visitor/message/domain/VisitorMessageService.java:37]: Move it to mapper!

**Review by misterrevo — COMMENTED** [review]
Close to end!! Good job, but fix what i comment soon, then like u said u will write tests

File: src/main/java/karuna/karuna_backend/visitor/message/domain/VisitorMessageService.java (hunk 8 above)
Comment (KrystynaCal) [reply src/main/java/karuna/karuna_backend/visitor/message/domain/VisitorMessageService.java:34]: I moved to mapper and right now in mapper  we can shorten message. I added new commits

Comment (KrystynaCal) [reply src/main/java/karuna/karuna_backend/visitor/message/domain/VisitorMessageService.java:34]: is this ok right now?

File: src/main/java/karuna/karuna_backend/security/config/WebSecurityConfig.java [hunk 9]
```diff
@@ -42,11 +43,17 @@ SecurityFilterChain filterChain(HttpSecurity http) throws Exception {
                                 "/swagger-ui/**",
                                 "/swagger-ui.html",
                                 "/v3/api-docs/**",
-                                "/swagger-resources/**",
-                                "/visitors/messages/**")
+                                "/swagger-resources/**")
                         .permitAll()
+                        .requestMatchers(HttpMethod.GET, "/visitors/messages").hasRole("ADMIN")
```
Comment (misterrevo) [inline src/main/java/karuna/karuna_backend/security/config/WebSecurityConfig.java:48]: Look how i refactore security, do the same

File: src/main/java/karuna/karuna_backend/visitor/message/domain/VisitorMessageService.java (hunk 8 above)
Comment (misterrevo) [reply src/main/java/karuna/karuna_backend/visitor/message/domain/VisitorMessageService.java:34]: Move createing of wrapper to mapper

File: src/main/java/karuna/karuna_backend/visitor/message/infrastructure/VisitorMessageController.java [hunk 10]
```diff
@@ -47,4 +51,21 @@ VisitorMessageDto sendMessage(@RequestBody @Valid VisitorMessageCreateDto visito
...
+            @ApiResponse(
+                    responseCode = "200",
+                    description = "List od messages",
+                    content = @Content(mediaType = "application/json", schema = @Schema(implementation = VisitorMessageWrapper.class))),
+            @ApiResponse(
+                    responseCode = "403",
+                    description = "Forbidden: You don't have permission to access this resource",
+                    content = @Content(mediaType = "application/json"))
+    })
+    @GetMapping
```
Comment (misterrevo) [inline src/main/java/karuna/karuna_backend/visitor/message/infrastructure/VisitorMessageController.java:67]: Protect it for be online for admin

File: src/main/java/karuna/karuna_backend/security/config/WebSecurityConfig.java [hunk 11]
```diff
@@ -64,3 +64,9 @@ private static AuthorizeHttpRequestsConfigurer<HttpSecurity>.AuthorizationManage
                 .permitAll();
     }
 }
+
```
Comment (misterrevo) [inline src/main/java/karuna/karuna_backend/security/config/WebSecurityConfig.java:67]: what is it?

Comment (KrystynaCal) [reply src/main/java/karuna/karuna_backend/security/config/WebSecurityConfig.java:67]: the same on master

File: src/main/java/karuna/karuna_backend/security/config/WebSecurityConfig.java [hunk 12]
```diff
@@ -48,6 +48,7 @@ private AuthorizeHttpRequestsConfigurer<HttpSecurity>.AuthorizationManagerReques
         return configurePublicHttpRequests(auth)
                 .requestMatchers(HttpMethod.GET, "/contents/**")
                 .permitAll()
+                .requestMatchers(HttpMethod.GET, "/visitors/messages").hasRole("ADMIN")
```
Comment (misterrevo) [inline src/main/java/karuna/karuna_backend/security/config/WebSecurityConfig.java:51]: No!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
In controller we have security right now!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

File: src/test/java/karuna/karuna_backend/visitor/message/domain/VisitorMessageServiceTest.java [hunk 13]
```diff
@@ -24,4 +30,21 @@ void shouldSendMessage() {
...
+
+
+    @Test
+    void shouldReceiveShortenMessage() {
+        //given
+        VisitorMessageCreateDto visitorMessageCreateDto = new VisitorMessageCreateDto(Constants.LONG_BODY, Constants.CONTACT);
+        VisitorMessageRequest visitorMessageRequest = new VisitorMessageRequest(0, 10, 20);
+        VisitorMessage visitorMessage = VisitorMessageMapper.toModel(visitorMessageCreateDto);
+        visitorMessageRepository = new InMemoryVisitorMessageRepository();
+        visitorMessageRepository.save(visitorMessage);
```
Comment (misterrevo) [inline src/test/java/karuna/karuna_backend/visitor/message/domain/VisitorMessageServiceTest.java:42]: Why u test by repository? Test by service like in other tests

File: src/main/java/karuna/karuna_backend/visitor/message/domain/VisitorMessageService.java [hunk 14]
```diff
@@ -19,4 +24,16 @@ public VisitorMessageDto sendMessage(VisitorMessageCreateDto visitorMessageCreat
         emailSender.sendEmail("HARDCODED-EMAIL@EMAIL.PL", "KARUN VISTOR MESSAGE", savedVisitorMessage.getBody(), null);
         return VisitorMessageMapper.toDto(savedVisitorMessage);
     }
+
+
+    public VisitorMessageWrapper getMessages(VisitorMessageRequest visitorMessageRequest) {
+        PageRequest pageRequest = PageRequest.of(visitorMessageRequest.offset(), visitorMessageRequest.limit());
```
Comment (misterrevo) [inline src/main/java/karuna/karuna_backend/visitor/message/domain/VisitorMessageService.java:30]: We duplicate some fields from pagerequest, what if pagerequst will be fragmnt of pagerequest? @BarSwi @KrystynaCal

File: src/test/java/karuna/karuna_backend/visitor/message/domain/InMemoryVisitorMessageRepository.java [hunk 15]
```diff
@@ -176,4 +179,15 @@ public List<VisitorMessage> findAll(Sort sort) {
     public Page<VisitorMessage> findAll(Pageable pageable) {
         return null;
     }
+
+    private List<VisitorMessage> messages = new ArrayList<>();
```
Comment (misterrevo) [inline src/test/java/karuna/karuna_backend/visitor/message/domain/InMemoryVisitorMessageRepository.java:183]: Why u not use database?

Comment (misterrevo) [reply src/test/java/karuna/karuna_backend/visitor/message/domain/InMemoryVisitorMessageRepository.java:183]: private HashMap<Long, VisitorMessage> database = new HashMap<>();

File: src/test/java/karuna/karuna_backend/visitor/message/domain/VisitorMessageServiceTest.java [hunk 16]
```diff
@@ -24,4 +29,18 @@ void shouldSendMessage() {
         assertEquals(Constants.BODY, visitorMessageDto.body());
         assertEquals(Constants.CONTACT, visitorMessageDto.contact());
     }
+
+
+    @Test
+    void shouldReceiveShortenMessage() {
+        //given
+        VisitorMessageCreateDto visitorMessageCreateDto = new VisitorMessageCreateDto(Constants.LONG_BODY, Constants.CONTACT);
```
Comment (misterrevo) [inline src/test/java/karuna/karuna_backend/visitor/message/domain/VisitorMessageServiceTest.java:37]: This is unused

File: src/test/java/karuna/karuna_backend/visitor/message/domain/VisitorMessageServiceTest.java [hunk 17]
```diff
@@ -24,4 +29,18 @@ void shouldSendMessage() {
...
+        VisitorMessageRequest visitorMessageRequest = new VisitorMessageRequest(0, 10, 20);
+        //when
+        VisitorMessageWrapper shortenedMessages = visitorMessageService.getMessages(visitorMessageRequest);
+        List<VisitorMessageDto> messages = shortenedMessages.messages();
+        //then
+        for (VisitorMessageDto message : messages)
```
Comment (misterrevo) [inline src/test/java/karuna/karuna_backend/visitor/message/domain/VisitorMessageServiceTest.java:43]: use brackets {} and this loop never run !! so its reason why test success

File: src/main/java/karuna/karuna_backend/visitor/message/domain/VisitorMessageService.java (hunk 14 above)
Comment (misterrevo) [reply src/main/java/karuna/karuna_backend/visitor/message/domain/VisitorMessageService.java:30]: @BarSwi what u think?

Comment (BarSwi) [reply src/main/java/karuna/karuna_backend/visitor/message/domain/VisitorMessageService.java:30]: What exactly do you mean by duplicate and "pagerequest will be fragment of pagerequest"

Comment (misterrevo) [reply src/main/java/karuna/karuna_backend/visitor/message/domain/VisitorMessageService.java:30]: I want to move pagerequest into visitormessagerequst, and remove offset and limit 🗡️  i write it chaotic, my bad

Comment (KrystynaCal) [reply src/main/java/karuna/karuna_backend/visitor/message/domain/VisitorMessageService.java]: I fix this. Hope is ok :)

File: src/main/java/karuna/karuna_backend/visitor/message/dto/VisitorMessageRequest.java [hunk 18]
```diff
@@ -0,0 +1,11 @@
+package karuna.karuna_backend.visitor.message.dto;
+
+import org.springframework.data.domain.PageRequest;
+
+public record VisitorMessageRequest(int offset,
+                                    int limit,
+                                    int bodyLenLimit) {
+    public PageRequest toPageRequest() {
```
Comment (misterrevo) [inline src/main/java/karuna/karuna_backend/visitor/message/dto/VisitorMessageRequest.java:8]: No!! 
public record VisitorMessageRequest(PageRequest request,
                                    int bodyLenLimit)  like this!

File: src/test/java/karuna/karuna_backend/visitor/message/domain/VisitorMessageServiceTest.java [hunk 19]
```diff
@@ -24,4 +31,20 @@ void shouldSendMessage() {
...
+        //given
+        VisitorMessageCreateDto visitorMessageCreateDto = new VisitorMessageCreateDto(Constants.LONG_BODY, Constants.CONTACT);
+        VisitorMessageDto visitorMessageDto = visitorMessageService.sendMessage(visitorMessageCreateDto);
+        VisitorMessageRequest visitorMessageRequest = new VisitorMessageRequest(PageRequest.of(0, 10, Sort.Direction.DESC, "createdAt"), 5);
+        //when
+        VisitorMessageWrapper shortenedMessages = visitorMessageService.getMessages(visitorMessageRequest);
+        List<VisitorMessageDto> messages = shortenedMessages.messages();
+        //then
+        for (VisitorMessageDto message : messages) {
+            assertEquals(5, message.body().length());
```
Comment (misterrevo) [inline src/test/java/karuna/karuna_backend/visitor/message/domain/VisitorMessageServiceTest.java:47]: Last cosmetic think, add before this loop assertion which check size of messages is greater than 1, becuase this loop can be not invoked and test give false success result

Comment (KrystynaCal) [reply src/test/java/karuna/karuna_backend/visitor/message/domain/VisitorMessageServiceTest.java:49]: corrected :)

**Review by misterrevo — APPROVED** [review]

**Pull request merged**